File: src/main/java/vserver/util/ByteArrayCharStream.java
package vserver.util;

import vjson.CharStream;
import vproxy.util.ByteArray;

public class ByteArrayCharStream implements CharStream {
    private final ByteArray array;
    private int idx = -1;

    public ByteArrayCharStream(ByteArray array) {
        this.array = array;
    }

    @Override
    public boolean hasNext(int i) {
        return idx + i < array.length();
    }

    @Override
    public char moveNextAndGet() {
        return (char) array.get(++idx);
    }

    @Override
    public char peekNext(int i) {
        return (char) array.get(idx + i);
    }
}

File: lib/src/main/kotlin/vproxy/lib/common/ByteArrayCharStream.kt
package vproxy.lib.common

import vjson.CharStream
import vproxy.base.util.ByteArray
import java.nio.charset.Charset

class ByteArrayCharStream(data: ByteArray, charset: Charset) : CharStream {
  private val chars: CharArray = String(data.toJavaArray(), charset).toCharArray()
  private var idx = -1
  override fun hasNext(i: Int): Boolean {
    return idx + i < chars.size
  }

  override fun moveNextAndGet(): Char {
    return chars[++idx]
  }

  override fun peekNext(i: Int): Char {
    return chars[idx + i]
  }
}
